### DISCUSSION TITLE
v8.0.0

### DISCUSSION THREAD

**Discussion by tagyoureit** [tagyoureit/nodejs-poolController#702] [Announcements]
v8.0.0
* Support for Screenlogic
* Refactor back end code
* Node v16/v18 support (eol for v12/v14)
* Easytouch v1 support
* Ansql-25 (mock controller)

<hr /><em>This discussion was created from the release <a href='https://github.com/tagyoureit/nodejs-poolController/releases/tag/v8.0.0'>v8.0.0</a>.</em>

**Comment by johnny2678** [comment]
Congrats on V8 @tagyoureit  and @rstrouse !

Excited to give it a shot but things have been humming along pretty well and I don't want to break my v7 setup.

Maybe just gzip the njspc* dirs and save them off so I can restore if needed?

I use N for node versioning so rolling back node should be easy

**Reply by Telemachus1977** [reply]
I agree. I am very not tempted to even try this on my main setup.

I made toy with it with my bench setup.

**Comment by arrmo** [comment]
Trying it here ... LOL. Getting a lot of this in the logs - expected?
```
1|njsPC  | CALLBACKED
1|njsPC  | CALLBACKED
1|njsPC  | CALLBACKED
1|njsPC  | CALLBACKED
```

And @rstrouse, does this include the `virtual circuit Spa Heater` fix we were playing with recently.

So far, seems to be up and running. Thanks!

**Comment by arrmo** [comment]
And, from the logs,
```
1|njsPC  |   == Node detected an unhandled rejection! ==
1|njsPC  |   Cannot read properties of undefined (reading 'setCircuitStateAsync')
1|njsPC  |   TypeError: Cannot read properties of undefined (reading 'setCircuitStateAsync')
1|njsPC  |       at SLCircuits.setCircuitStateAsync (/home/rmorris/Documents/nodejs-poolController/controller/comms/ScreenLogic.ts:1224:33)
1|njsPC  |       at ITTouchCircuitCommands.setCircuitStateAsync (/home/rmorris/Documents/nodejs-poolController/controller/boards/EasyTouchBoard.ts:1535:31)
1|njsPC  |       at Object.eval [as fnProcessor] (eval at bindProcessor (/home/rmorris/Documents/nodejs-poolController/web/interfaces/ruleInterface.ts:36:39), <anonymous>:10:20)
1|njsPC  |       at RuleInterfaceBindings.executeProcessor (/home/rmorris/Documents/nodejs-poolController/web/interfaces/ruleInterface.ts:46:55)
1|njsPC  |       at RuleInterfaceBindings.bindEvent (/home/rmorris/Documents/nodejs-poolController/web/interfaces/ruleInterface.ts:69:26)
1|njsPC  |       at RuleInterfaceServer.emitToClients (/home/rmorris/Documents/nodejs-poolController/web/Server.ts:1234:27)
1|njsPC  |       at WebServer.emitToClients (/home/rmorris/Documents/nodejs-poolController/web/Server.ts:148:30)
1|njsPC  |       at TemperatureState.emitData (/home/rmorris/Documents/nodejs-poolController/controller/State.ts:557:55)
1|njsPC  |       at TemperatureState.emitEquipmentChange (/home/rmorris/Documents/nodejs-poolController/controller/State.ts:551:39)
1|njsPC  |       at DirtyStateCollection.emitChanges (/home/rmorris/Documents/nodejs-poolController/controller/State.ts:725:20)
1|njsPC  |       at State.emitEquipmentChanges (/home/rmorris/Documents/nodejs-poolController/controller/State.ts:275:74)
1|njsPC  |       at Function.processTouchCircuits (/home/rmorris/Documents/nodejs-poolController/controller/comms/messages/status/EquipmentStateMessage.ts:761:15)
1|njsPC  |       at Function.process (/home/rmorris/Documents/nodejs-poolController/controller/comms/messages/status/EquipmentStateMessage.ts:498:55)
1|njsPC  |       at Inbound.processBroadcast (/home/rmorris/Documents/nodejs-poolController/controller/comms/messages/Messages.ts:687:47)
1|njsPC  |       at Inbound.process (/home/rmorris/Documents/nodejs-poolController/controller/comms/messages/Messages.ts:790:22)
1|njsPC  |       at RS485Port.processCompletedMessage (/home/rmorris/Documents/nodejs-poolController/controller/comms/Comms.ts:1143:17)
```

**Comment by bmn001** [comment]
This is a fresh install for me.  Node 16.19.0 on a Raspberry Pi Zero W.

I'm also seeing a lot of CALLBACKED in the logs, then an eventual crash:

```
/home/pi/nodejs-poolController/controller/Equipment.ts:656
        let r = extend(true, { start: 1, end: 255 }, {start: range.start }, {end: range.end});  
                                                                   ^
TypeError: Cannot read properties of undefined (reading 'start')
    at PumpCollection.getNextEquipmentId (/home/pi/nodejs-poolController/controller/Equipment.ts:656:68)
    at Function.processIntelliFlo4 (/home/pi/nodejs-poolController/controller/comms/messages/config/RemoteMessage.ts:176:35)
    at Function.processRemote_IT (/home/pi/nodejs-poolController/controller/comms/messages/config/RemoteMessage.ts:151:35)
    at Function.process (/home/pi/nodejs-poolController/controller/comms/messages/config/RemoteMessage.ts:57:27)
    at Inbound.processBroadcast (/home/pi/nodejs-poolController/controller/comms/messages/Messages.ts:730:39)
    at Inbound.process (/home/pi/nodejs-poolController/controller/comms/messages/Messages.ts:790:22)
    at RS485Port.processCompletedMessage (/home/pi/nodejs-poolController/controller/comms/Comms.ts:1143:17)
    at RS485Port.processInboundPackets (/home/pi/nodejs-poolController/controller/comms/Comms.ts:1184:48)
    at RS485Port.processPackets (/home/pi/nodejs-poolController/controller/comms/Comms.ts:935:14)
    at Immediate.<anonymous> (/home/pi/nodejs-poolController/controller/comms/Comms.ts:896:114)
    at processImmediate (node:internal/timers:466:21)

```

**Reply by rstrouse** [reply]
Pull njsPC.  I fixed the issue with the equipment ranges when an IntelliFlo4 is involved.

@armo - The code should be there for the spa heater.  I do not see any dangling updates.

**Reply by tagyoureit** [reply]
My power has been out for 48 hours and could be out another 48. Great time to do a new release, lol!  

Thanks @rstrouse for patching some holes.

**Reply by jrhubott** [reply]
I pulled the latest code along with the change this update.  I'm also seeing the same issue as njsPC.  This is related to setCircuitStateAsync not getNextEquipmentId that was fixed in the last push

`
3/15/2023, 7:22:28 PM] info: [7:22:28 PM] 192.168.1.8 PUT /state/circuit/setState {"id":2,"state":true}
[3/15/2023, 7:22:28 PM] error: Cannot read properties of undefined (reading 'setCircuitStateAsync')
[3/15/2023, 7:22:40 PM] info: [7:22:40 PM] 192.168.1.8 PUT /state/circuit/setState {"id":4,"state":true}
[3/15/2023, 7:22:40 PM] error: Cannot read properties of undefined (reading 'setCircuitStateAsync')
`

**Reply by rstrouse** [reply]
What board are you running. Nixie, IntelliCenter, or one of the *Touch boards.

**Reply by jrhubott** [reply]
Easytouch.

**Reply by jrhubott** [reply]
I dug into it but not really sure what is going on.  setCircuitStateAsync is defined and implemented in [controller/boards/EasyTouchBoard.ts](https://github.com/tagyoureit/nodejspoolController/blob/88412a0fdaa8caaa718183e8c9e0882633471065/controller/boards/EasyTouchBoard.ts) so I'm not really sure what is going on,

**Reply by rstrouse** [reply]
Can I get you to create a replay for me?  That way I can model your system to see where it is going awry.

https://github.com/tagyoureit/nodejs-poolController/wiki/How-to-capture-all-packets-for-issue-resolution

**Reply by jrhubott** [reply]
I attempted to do a replay but now I'm stuck in a reboot loop with the following.  I attached the replay of it also.
<
[3/16/2023, 9:58:48 AM] verbose: Wrote packet [Port 0 id: 139] [255,0,255,165,1,16,33,214,1,0,1,174].Retries remaining: 3 
/app/controller/Equipment.ts:1401
    public getNextEquipmentId(range?: EquipmentIdRange, exclude?: number[]): number { return this.getNextEquipmentId(typeof range === 'undefined' ? sys.board.equipmentIds.pumps : range, exclude); }
                                                                                                  ^
RangeError: Maximum call stack size exceeded
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
    at PumpCollection.getNextEquipmentId (/app/controller/Equipment.ts:1401:99)
Init state for Pool Controller
>
[replay (1).zip](https://github.com/tagyoureit/nodejs-poolController/files/10993638/replay.1.zip)

**Reply by bmn001** [reply]
I pulled the latest and now am getting the exact [same error](https://github.com/tagyoureit/nodejs-poolController/discussions/702#discussioncomment-5336890) as @jrhubott.

I've got a RS485 adapter connected and an EasyTouch 8/Screenlogic controller on my network.

My console and packet logs: [logs.zip](https://github.com/tagyoureit/nodejs-poolController/files/10993826/logs.zip)

**Reply by arrmo** [reply]
Dumb question (likely me) ... but it seems like the updates are on the `master` branch, not `v8.0.0` ... do I have that right? So pull master?

Thanks!

**Reply by rstrouse** [reply]
Yes pull master.  The 8.0.0 branch will eventually be deprecated.

**Comment by arrmo** [comment]
OK, did - guessed right 🤣. Getting (in the logs - and a popup in the browser matching this),
```
1|njsPC  | [3/16/2023, 2:22:58 PM] error: Cannot read properties of undefined (reading 'setCircuitStateAsync')
1|njsPC  | unhandled rejection
1|njsPC  |   == Node detected an unhandled rejection! ==
1|njsPC  |   Cannot read properties of undefined (reading 'setCircuitStateAsync')
1|njsPC  |   TypeError: Cannot read properties of undefined (reading 'setCircuitStateAsync')
1|njsPC  |       at SLCircuits.setCircuitStateAsync (/home/rmorris/Documents/nodejs-poolController/controller/comms/ScreenLogic.ts:1224:33)
1|njsPC  |       at ITTouchCircuitCommands.setCircuitStateAsync (/home/rmorris/Documents/nodejs-poolController/controller/boards/EasyTouchBoard.ts:1535:31)
1|njsPC  |       at Object.eval [as fnProcessor] (eval at bindProcessor (/home/rmorris/Documents/nodejs-poolController/web/interfaces/ruleInterface.ts:36:39), <anonymous>:10:20)
1|njsPC  |       at RuleInterfaceBindings.executeProcessor (/home/rmorris/Documents/nodejs-poolController/web/interfaces/ruleInterface.ts:46:55)
1|njsPC  |       at RuleInterfaceBindings.bindEvent (/home/rmorris/Documents/nodejs-poolController/web/interfaces/ruleInterface.ts:69:26)
1|njsPC  |       at RuleInterfaceServer.emitToClients (/home/rmorris/Documents/nodejs-poolController/web/Server.ts:1234:27)
1|njsPC  |       at WebServer.emitToClients (/home/rmorris/Documents/nodejs-poolController/web/Server.ts:148:30)
1|njsPC  |       at TemperatureState.emitData (/home/rmorris/Documents/nodejs-poolController/controller/State.ts:557:55)
1|njsPC  |       at TemperatureState.emitEquipmentChange (/home/rmorris/Documents/nodejs-poolController/controller/State.ts:551:39)
1|njsPC  |       at DirtyStateCollection.emitChanges (/home/rmorris/Documents/nodejs-poolController/controller/State.ts:725:20)
1|njsPC  |       at State.emitEquipmentChanges (/home/rmorris/Documents/nodejs-poolController/controller/State.ts:275:74)
1|njsPC  |       at Function.processTouchCircuits (/home/rmorris/Documents/nodejs-poolController/controller/comms/messages/status/EquipmentStateMessage.ts:761:15)
1|njsPC  |       at Function.process (/home/rmorris/Documents/nodejs-poolController/controller/comms/messages/status/EquipmentStateMessage.ts:498:55)
1|njsPC  |       at Inbound.processBroadcast (/home/rmorris/Documents/nodejs-poolController/controller/comms/messages/Messages.ts:687:47)
1|njsPC  |       at Inbound.process (/home/rmorris/Documents/nodejs-poolController/controller/comms/messages/Messages.ts:790:22)
1|njsPC  |       at RS485Port.processCompletedMessage (/home/rmorris/Documents/nodejs-poolController/controller/comms/Comms.ts:1143:17)
```

I'm on Node.js v18, did an `npm i` of course. Thoughts?

Thanks!

**Reply by rstrouse** [reply]
This indicates that you have a port assigned to ScreenLogic.  The code looks to see if a port is assigned and it calls out to ScreenLogic instead of RS485.  Since SL is not awake on your connection it throws this error.  I need somebody to give me a full replay to replicate this though.

**Reply by jrhubott** [reply]
Same thing.  When I try to perform a replay it switches my controller to Nixie and drops the connection.  I restarted the pool controller with capture still on though so hopefully this contains the correct replay.

The master branch fixed the recursive loop at startup.
[replay (2).zip](https://github.com/tagyoureit/nodejs-poolController/files/10995532/replay.2.zip)

I don't have ScreenLogic assigned anywhere only RS232 to a 2 body EasyTouch8.

**Reply by rstrouse** [reply]
Ok go into the config menu in dashPanel then open the Controller --> Comms tab. Expand the primary port and hit save.  This should force the comms type to either network or local.

**Reply by jrhubott** [reply]
I just look at my config.  It does have a screenlogic connection configure as local.  I removed it from the configuration but when I start/restart the pool controller the section comes back.

**Reply by rstrouse** [reply]
Yes it will put that back you need to save the port like I said and it will add an element in there for comms that says `"type": "local",

**Reply by jrhubott** [reply]
Just saw your message.  That worked.

**Reply by rstrouse** [reply]
Btw thanks for the replay.  I would not have figured it out without it.

**Reply by jrhubott** [reply]
No problem.  I really appreciate the time you spent on this.  I also had to add multer to the docker build of the dashpanel.  Without adding it to the dist image it would fail on startup.  Not sure why as it was listed in the package.json

**Reply by arrmo** [reply]
> Ok go into the config menu in dashPanel then open the Controller --> Comms tab. Expand the primary port and hit save. This should force the comms type to either network or local.

Did the same thing here, no changes at all ... and now it seems OK. Does that make sense? Did some settings get messed up on upgrade somehow?

Thanks!

**Reply by rstrouse** [reply]
Yes, originally it assumed a local connection.  However, when the ScreenLogic connection was added it assumed it was the default.  Which of course re-routed the traffic to a non-existed port.

**Reply by arrmo** [reply]
Got it, thanks!
